xdp-dump/xdpdump.c

```diff
@@ -0,0 +1,628 @@
+	str[i] = 0;
+	return 0;
+}
+
+/*****************************************************************************
+ * handle_perf_event()
+ *****************************************************************************/
+static enum bpf_perf_event_ret handle_perf_event(void *private_data,
+						 int cpu,
+						 struct perf_event_header *event)
+{
+	uint64_t ts;
+	struct perf_handler_ctx *ctx = private_data;
+	struct perf_sample_event *e = container_of(event,
+						   struct perf_sample_event,
+						   header);
+	struct perf_lost_event *lost = container_of(event,
+						    struct perf_lost_event,
+						    header);
+
+	switch(e->header.type) {
+	case PERF_RECORD_SAMPLE:
+
+		if (e->header.size < sizeof(struct perf_sample_event) ||
+		    e->size < (sizeof(struct pkt_trace_metadata) + e->metadata.cap_len))
+			return LIBBPF_PERF_EVENT_CONT;
+
+		ts = e->time + ctx->epoch_delta;
+		if (ctx->pcap_dumper) {
+			struct pcap_pkthdr h;
+
+			h.ts.tv_sec = ts / 1000000000ULL;
+			h.ts.tv_usec = ts % 1000000000ULL / 1000;
+			h.caplen = min(e->metadata.cap_len, ctx->cfg->snaplen);
+			h.len = e->metadata.pkt_len;
+			pcap_dump((u_char *) ctx->pcap_dumper, &h,
+				  e->packet);
+		} else {
+			int i;
+			char hline[SNPRINTH_MIN_BUFFER_SIZE];
+			bool fexit = e->metadata.flags & MDF_DIRECTION_FEXIT;
+
+			printf("%llu.%09lld: packet size %u bytes, "
+			       "captured %u bytes on "
+			       "if_index %u, rx queue %u, @%s\n",
+			       ts / 1000000000ULL,
+			       ts % 1000000000ULL,
+			       e->metadata.pkt_len,
+			       e->metadata.cap_len, e->metadata.ifindex,
+			       e->metadata.rx_queue,
+			       fexit ? "fexit" : "fentry");
+
+			for (i = 0; i < e->metadata.cap_len; i += 16) {
+				snprinth(hline, sizeof(hline),
+					 e->packet, e->metadata.cap_len, i);
+				printf("  %s\n", hline);
+			}
+		}
+		ctx->captured_packets++;
+		break;
+
+	case PERF_RECORD_LOST:
+		ctx->missed_events += lost->lost;
+		break;
+	}
+
+	return LIBBPF_PERF_EVENT_CONT;
+}
+
+/*****************************************************************************
+ * get_epoch_to_uptime_delta()
+ *****************************************************************************/
+static uint64_t get_epoch_to_uptime_delta(void)
+{
+	/* This function will calculate the rough delta between uptime
+	 * seconds and the epoch time. This is not a precise delta, but it's
+	 * good enough to get a general offset. The delta between packets is
+	 * still based on the timestamps from the trace infrastructure.
+	 */
+	struct timespec ts;
+	uint64_t        uptime;
+	uint64_t        epoch = time(NULL) * 1000000000ULL;
+
+	clock_gettime(CLOCK_MONOTONIC, &ts);
+	uptime = ts.tv_sec * 1000000000ULL + ts.tv_nsec;
+
+	return epoch - uptime;
+}
+
+/*****************************************************************************
+ * capture_on_legacy_interface()
+ *****************************************************************************/
+static bool capture_on_legacy_interface(struct dumpopt *cfg)
```

Comment:
I can see why it may be useful to have tool not fail when there's no XDP program loaded, but I wonder if it's not better to just load one? :)

Reply:
Thought about this, but it might alter the behaviour of the system. So I decided to just fall back to normal operation and print a warning. The goal for this was that a user could use the same tool and work as expected.

Reply:
Yeah, fair point, loading an XDP program can have side effects... Hmm, I wonder if we should just bail then? My fear is that we'll end up re-implementing a half-baked tcpdump if people start using this feature, basically... :)

Reply:
Had this discussion with Jiri also, he feels its better to warn and continue as people can do what they want to do using a single tool.

Reply:
Right, well, as long as "what they want to do" does not end up requiring a full re-implementation of tcpdump...

I guess we can keep it, but just be warned that I plan to push back on any feature creep in this area :)

Reply:
I agree it should only allow for the same functionality as the XDP trace, no additional items in this area!